xilem/src/driver.rs

```diff
@@ -1,67 +1,125 @@
 // Copyright 2024 the Xilem Authors
 // SPDX-License-Identifier: Apache-2.0
 
-use masonry::{app_driver::AppDriver, widget::RootWidget};
-use xilem_core::MessageResult;
+use std::sync::Arc;
+
+use masonry::{
+    app_driver::AppDriver,
+    event_loop_runner::{self, EventLoopProxy, MasonryUserEvent},
+    widget::RootWidget,
+    WidgetId,
+};
+use xilem_core::{DynMessage, Message, MessageResult, ProxyError, RawProxy, ViewId};
 
 use crate::{ViewCtx, WidgetView};
 
 pub struct MasonryDriver<State, Logic, View, ViewState> {
     pub(crate) state: State,
     pub(crate) logic: Logic,
     pub(crate) current_view: View,
-    pub(crate) view_ctx: ViewCtx,
+    pub(crate) ctx: ViewCtx,
     pub(crate) view_state: ViewState,
 }
 
+/// The `WidgetId` which async events should be sent to.
+pub const ASYNC_MARKER_WIDGET: WidgetId = WidgetId::reserved(0x1000);
+
+/// The action which should be used for async events.
+pub fn async_action(path: Arc<[ViewId]>, message: Box<dyn Message>) -> masonry::Action {
+    masonry::Action::Other(Box::<MessagePackage>::new((path, message)))
+}
+
+/// The type used to send a message for async events.
+type MessagePackage = (Arc<[ViewId]>, DynMessage);
+
+impl RawProxy for MasonryProxy {
+    fn send_message(&self, path: Arc<[ViewId]>, message: DynMessage) -> Result<(), ProxyError> {
+        match self
+            .0
+            .send_event(event_loop_runner::MasonryUserEvent::Action(
+                async_action(path, message),
+                ASYNC_MARKER_WIDGET,
+            )) {
+            Ok(()) => Ok(()),
+            Err(err) => {
+                let MasonryUserEvent::Action(masonry::Action::Other(res), _) = err.0 else {
+                    unreachable!(
+                        "We know this is the value we just created, which matches this pattern"
+                    )
+                };
+                Err(ProxyError::DriverFinished(
+                    res.downcast::<MessagePackage>().unwrap().1,
+                ))
+            }
+        }
+    }
+}
+
+pub struct MasonryProxy(pub(crate) EventLoopProxy);
```

Comment:
I'm not sure if we should "abuse" `EventLoopProxy`, and whether we should have our "own" async event handling mechanism in xilem to have more control over the whole (async) data flow and view rendering, which combines either `MasonryDriver::on_action` as well as async handling of xilem, which would also make (most) of the changes in masonry unnecessary.

I'm mostly thinking about the (unfortunately a little bit complex) logic [here](https://github.com/linebender/xilem/blob/xilem_classic_archive/src/app.rs#L324-L387), which is somewhat explained [here](https://github.com/raphlinus/raphlinus.github.io/issues/74) (paragraph starting with "Ideal is hybrid").

I don't want to block on this, as this could as well be done in a follow-up PR, just something to think about...

Reply:
Needing to hard-code such a delay is one of those problems which goes away if we properly implement frame pacing. Actually doing that is of course a hard problem, but it's not an impossible one.

It is possible to move `app_logic`, etc. off of the main thread. However, I don't expect the cost of `app_logic` to be very high, so I strongly suspect it's not worth doing.

I'm afraid I don't really understand the rest of the suggestion you're making - we need some way to communicate with the main thread, and that mechanism is `EventLoopProxy`.

Reply:
Ah I think I get what you mean, i.e. we need access to the root widget, which is only possible in `on_action`. Ok, well then forget about it, I also think, that the rest could of course be done with custom messages only targeted for the driver impl (like it is already with `ASYNC_MARKER_WIDGET`), but as you say that may be done with frame pacing properly.